Sanitize CMake + directory structure, add Travis and AppVeyor CI

cernoch opened the pull request.

cernoch committed:
Travis and AppVeyor CI scripts added

cernoch committed:
Windows/MSVC compatibility

cernoch committed:
PBLib::ld64 is not GCC-dependent

cernoch committed:
Sanitize CMake (by rewriting from scratch)
- MiniSAT becomes a Git submodule.
- The BasicPBSolver.cc is extracted from
  the (now deleted) minisat.tgz archive.

horenmar commented:
Github somehow ate my comment, but note that the change in `PBLib::ld64` will cause a significant slowdown as compared to using the intrinsics. I would suggest using intrinsics on the big three compilers (MSVC's is called [BitScanReverse](https://docs.microsoft.com/en-us/cpp/intrinsics/bitscanreverse-bitscanreverse64), Clang should be able to use the same one as GCC), with a fallback to the loop on unknown compiler (together with `#pragma warn`).

cernoch commented:
I've noted the `PBlib::ld64` comment and put it to the TODO list.

cernoch commented:
On OS-X, there is a [strange issue](https://travis-ci.org/master-keying/pblib/jobs/380596409#L139):
```
ld: warning: directory not found for option '-L/Users/travis/build/master-keying/pblib/BasicPBSolver/minisat'
Undefined symbols for architecture x86_64:
  "___cxa_deleted_virtual", referenced from:
      vtable for AuxVarManager in libpblib.a(auxvarmanager.cpp.o)
      vtable for PB2CNF in libpblib.a(pb2cnf.cpp.o)
      vtable for PreEncoder in libpblib.a(preencoder.cpp.o)
      vtable for IncSimplePBConstraint in libpblib.a(IncSimplePBConstraint.cpp.o)
```

Any idea how to fix this?

horenmar commented:
There is a good chance that the OS X image uses old libcxx, [because as far as I can tell this was patched in late 2017](https://reviews.llvm.org/D37830) (and I have no idea how often is the underlying libcxx updated on OS X).


----

e: a simple, but temporary! fix is to export a simple function like so:

```cpp
extern "C" void __cxa_deleted_virtual() { 
    abort();
}
```

This is pretty much what the function is supposed to do, and will work until there is a link conflict because of updated libcxx. Fancy solution would be to also decorate the function with weak linkage.

cernoch committed:
Hotfix: Travis can build the source on OS-X

cernoch commented:
Now I consider the patchset ready.

Are there any more comments or issues?

If not, I will rebase-merge the commits.

horenmar commented:
Doing `make -j` in TravisCI script will decrease build performance relative to `make -j 2`. Ideally we could discover the number of cores available at runtime, but Travis does not provide a good way to do so (the VMs report as having 8+ core CPUs, but only 2 are provided to any given build). In practice, the ideal number of parallel builds can be different from number of cores, e.g. to avoid IO bottlenecks, but using the same number of parallel jobs as there are available CPUs is a good default.


------
Quick notes:
* The include paths will not work properly after an install
* Globs in CMakeLists.txt run into problem where you need to tell CMake manually to redo configuration step for it to discover new files.

I will expand upon this tomorrow.

horenmar commented on BasicPBSolver/BasicSATSolver.cc:
```diff
@@ -0,0 +1,155 @@
+
+
+
+#include "BasicSATSolver.h"
```
This does not seem right. Also, I have no idea where this file came from, as I don't see any corresponding deletions in the changeset.

horenmar commented on BasicPBSolver/BasicSATSolver.cc:
```diff
@@ -0,0 +1,155 @@
+}
+
+
+BasicSATSolver::BasicSATSolver()
```
This and the destructor should be just `= default;`ed in the header file.

horenmar commented on CMakeLists.txt:
```diff
@@ -1,63 +1,46 @@
 PROJECT(pblib)
```
If we do `project(pblib LANGUAGES CXX)`, cmake will not look for C compiler and configuring a project will be faster.

horenmar commented on CMakeLists.txt:
```diff
@@ -1,63 +1,46 @@
 PROJECT(pblib)
 
-cmake_minimum_required(VERSION 2.8)
```
This directive should _always_ be first thing in a CMakeLists.txt, see [documentation](https://cmake.org/cmake/help/latest/command/cmake_minimum_required.html)

horenmar commented on CMakeLists.txt:
```diff
@@ -1,63 +1,46 @@
+add_subdirectory(BasicPBSolver/minisat EXCLUDE_FROM_ALL)
 
+# pblib
+file(GLOB pblib_src_files "*.cpp" "encoder/*.cpp")
```
GLOB is not supported in all CMake targets, so this will be done at configuration time and will not be redone if new files are added. The files should be listed explicitly instead.

horenmar commented on encoder/adderencoding.h:
```diff
@@ -14,6 +14,7 @@
 namespace PBLib
 {
 int ld64(const uint64_t x);
+int ld64universal(const uint64_t x);
```
Isn't this method just an implementation detail of `ld64`? If it is, it should have internal linkage inside the implementation file and should not be visible from the outside.

horenmar reviewed (changes requested):
The `cerr`/`cout` printing thing is terrible, but not part of your code, so I didn't give it an inline comment. I would however suggest a change, so that the various `print` methods to take `std::ostream&` for output, and just write into it. Then a top-level class can provide the proper output stream as desired.

-------

CMakeLists.txt could be further improved -- currently it doesn't know how to install itself properly and it will not generate nice project file for IDEs. [This post](https://codingnest.com/basic-cmake/#generatedproject) provides some advice with the second. For the first, MiniSat can be copied mostly blindly. Just note that proper installation might require some changes to include paths.

horenmar committed:
Minor sanitization of C++ files
- Add missing <cmath> includes
- BasicSATSolver.h is included only once in .cpp
- BasicSATSolver has con/de-structors in the header file

cernoch committed:
Introduce "stuttering" directory structure and fix includes

Client code should #include <pblib/...> in all use-cases:
- Building an existing binary (pbencoder, pbo2maxsat, ...)
  both inside this repository as well as standalone files.
- Linking to pblib both in-source and as a system library.
- Building a custom binary without CMake (after
  having installed pblib as a system library).

cernoch commented:
Hopefully, all proposals have been addressed. @horenmar, can you check this, please?

horenmar commented:
Looks good, some extra stuff found:
* The CMake should write out PBLibConfig.cmake. It is possible to just write it out straight from the CMakeLists, or do like our MiniSat that has a trivial CMake .in file that is translated into the result.
* The anonymous namespace for generic clz is used wrong. It should not be mentioned in the header at all, and the definition inside `adderencoding.cpp` should be enclosed in the anonymous namespace instead.
